File: src/rssService.ts
import { RSSFeed, RSSArticle } from './types';
import { addArticles, getArticlesByFeed, updateFeed } from './db';

// CORS proxy for fetching RSS feeds
const CORS_PROXY = 'https://api.allorigins.win/raw?url=';

interface ParsedFeed {
  title: string;
  description?: string;
  link?: string;
  items: ParsedItem[];
}

interface ParsedItem {
  title: string;
  link: string;
  description?: string;
  content?: string;
  pubDate?: string;
  author?: string;
  guid?: string;
}

function parseXML(xmlString: string): Document {
  const parser = new DOMParser();
  return parser.parseFromString(xmlString, 'text/xml');
}

function getTextContent(element: Element | null): string {
  return element?.textContent?.trim() || '';
}

function parseRSSFeed(xml: Document): ParsedFeed {
  const channel = xml.querySelector('channel');
  if (!channel) {
    throw new Error('Invalid RSS feed: no channel element found');
  }

  const title = getTextContent(channel.querySelector('title'));
  const description = getTextContent(channel.querySelector('description'));
  const link = getTextContent(channel.querySelector('link'));

  const items: ParsedItem[] = [];
  const itemElements = channel.querySelectorAll('item');

  itemElements.forEach((item) => {
    const parsedItem: ParsedItem = {
      title: getTextContent(item.querySelector('title')),
      link: getTextContent(item.querySelector('link')),
      description: getTextContent(item.querySelector('description')),
      content: getTextContent(item.querySelector('content\\:encoded')) || 
               getTextContent(item.querySelector('content')),
      pubDate: getTextContent(item.querySelector('pubDate')),
      author: getTextContent(item.querySelector('author')) || 
              getTextContent(item.querySelector('dc\\:creator')),
      guid: getTextContent(item.querySelector('guid')) || 
            getTextContent(item.querySelector('link')),
    };
    items.push(parsedItem);
  });

  return { title, description, link, items };
}

function parseAtomFeed(xml: Document): ParsedFeed {
  const feed = xml.querySelector('feed');
  if (!feed) {
    throw new Error('Invalid Atom feed: no feed element found');
  }

  const title = getTextContent(feed.querySelector('title'));
  const subtitle = getTextContent(feed.querySelector('subtitle'));
  const linkElement = feed.querySelector('link[rel="alternate"]') || feed.querySelector('link');
  const link = linkElement?.getAttribute('href') || '';

  const items: ParsedItem[] = [];
  const entryElements = feed.querySelectorAll('entry');

  entryElements.forEach((entry) => {
    const entryLink = entry.querySelector('link[rel="alternate"]') || entry.querySelector('link');
    const parsedItem: ParsedItem = {
      title: getTextContent(entry.querySelector('title')),
      link: entryLink?.getAttribute('href') || '',
      description: getTextContent(entry.querySelector('summary')),
      content: getTextContent(entry.querySelector('content')),
      pubDate: getTextContent(entry.querySelector('published')) || 
               getTextContent(entry.querySelector('updated')),
      author: getTextContent(entry.querySelector('author name')),
      guid: getTextContent(entry.querySelector('id')) || 
            (entryLink?.getAttribute('href') || ''),
    };
    items.push(parsedItem);
  });

  return { title, description: subtitle, link, items };
}

export async function fetchFeed(url: string): Promise<ParsedFeed> {
  try {
    const response = await fetch(CORS_PROXY + encodeURIComponent(url));
    
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }

    const text = await response.text();
    const xml = parseXML(text);

    // Check for parsing errors
    const parserError = xml.querySelector('parsererror');
    if (parserError) {
      throw new Error('Failed to parse XML: ' + parserError.textContent);
    }

    // Detect feed type and parse accordingly
    if (xml.querySelector('rss') || xml.querySelector('channel')) {
      return parseRSSFeed(xml);
    } else if (xml.querySelector('feed')) {
      return parseAtomFeed(xml);
    } else {
      throw new Error('Unknown feed format');
    }
  } catch (error) {
    console.error('Error fetching feed:', error);
    throw error;
  }
}

export async function refreshFeed(feed: RSSFeed): Promise<number> {
  try {
    const parsedFeed = await fetchFeed(feed.url);
    
    // Update feed metadata
    const updatedFeed: RSSFeed = {
      ...feed,
      title: parsedFeed.title || feed.title,
      description: parsedFeed.description || feed.description,
      link: parsedFeed.link || feed.link,
      lastFetched: new Date(),
    };
    await updateFeed(updatedFeed);

    // Get existing articles for this feed
    const existingArticles = await getArticlesByFeed(feed.id);
    const existingGuids = new Set(existingArticles.map(a => a.guid));

    // Filter out articles that already exist
    const newArticles: RSSArticle[] = parsedFeed.items
      .filter(item => !existingGuids.has(item.guid || item.link))
      .map(item => ({
        id: crypto.randomUUID(),
        feedId: feed.id,
        title: item.title,
        link: item.link,
        description: item.description,
        content: item.content,
        pubDate: item.pubDate ? new Date(item.pubDate) : undefined,
        author: item.author,
        guid: item.guid || item.link,
        read: false,
        starred: false,
        createdAt: new Date(),
      }));

    if (newArticles.length > 0) {
      await addArticles(newArticles);
    }

    return newArticles.length;
  } catch (error) {
    console.error(`Error refreshing feed ${feed.title}:`, error);
    throw error;
  }
}

export async function refreshAllFeeds(feeds: RSSFeed[]): Promise<{ total: number; errors: string[] }> {
  let total = 0;
  const errors: string[] = [];

  for (const feed of feeds) {
    try {
      const newCount = await refreshFeed(feed);
      total += newCount;
    } catch (error) {
      errors.push(`${feed.title}: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  return { total, errors };
}

File: mobile/lib/services/rss_service.dart
import 'package:http/http.dart' as http;
import 'package:xml/xml.dart' as xml;

import '../models/rss_models.dart';
import 'db_service.dart';

class ParsedFeed {
  final String title;
  final String? description;
  final String? link;
  final List<ParsedItem> items;

  ParsedFeed({
    required this.title,
    this.description,
    this.link,
    required this.items,
  });
}

class ParsedItem {
  final String title;
  final String link;
  final String? description;
  final String? content;
  final String? pubDate;
  final String? author;
  final String? guid;
  final String? imageUrl;

  ParsedItem({
    required this.title,
    required this.link,
    this.description,
    this.content,
    this.pubDate,
    this.author,
    this.guid,
    this.imageUrl,
  });
}

class RefreshAllResult {
  final int total;
  final List<String> errors;

  RefreshAllResult({
    required this.total,
    required this.errors,
  });
}

class RssService {
  final DbService dbService;

  RssService({required this.dbService});

  Future<ParsedFeed> fetchFeed(String url) async {
    final response = await http.get(Uri.parse(url));
    if (response.statusCode != 200) {
      throw Exception('HTTP error ${response.statusCode}');
    }

    final document = xml.XmlDocument.parse(response.body);

    if (document.findAllElements('rss').isNotEmpty ||
        document.findAllElements('channel').isNotEmpty) {
      return _parseRss(document);
    } else if (document.findAllElements('feed').isNotEmpty) {
      return _parseAtom(document);
    } else {
      throw Exception('Unknown feed format');
    }
  }

  ParsedFeed _parseRss(xml.XmlDocument doc) {
    final channel = doc.findAllElements('channel').firstOrNull;
    if (channel == null) {
      throw Exception('Invalid RSS feed: no channel element found');
    }

    final title = _getText(channel.findElements('title').firstOrNull);
    final description =
        _getText(channel.findElements('description').firstOrNull);
    final link = _getText(channel.findElements('link').firstOrNull);

    final items = <ParsedItem>[];
    for (final item in channel.findElements('item')) {
      final parsedItem = ParsedItem(
        title: _getText(item.findElements('title').firstOrNull),
        link: _getText(item.findElements('link').firstOrNull),
        description: _getText(item.findElements('description').firstOrNull),
        content: _getText(_firstOf(item, ['content:encoded', 'content'])),
        pubDate: _getText(item.findElements('pubDate').firstOrNull),
        author:
            _getText(_firstOf(item, ['author', 'dc:creator'])),
        guid: _getText(item.findElements('guid').firstOrNull) ??
            _getText(item.findElements('link').firstOrNull),
        imageUrl: _extractImageUrl(item),
      );
      items.add(parsedItem);
    }

    return ParsedFeed(
      title: title,
      description: description,
      link: link,
      items: items,
    );
  }

  ParsedFeed _parseAtom(xml.XmlDocument doc) {
    final feed = doc.findAllElements('feed').firstOrNull;
    if (feed == null) {
      throw Exception('Invalid Atom feed: no feed element found');
    }

    final title = _getText(feed.findElements('title').firstOrNull);
    final subtitle = _getText(feed.findElements('subtitle').firstOrNull);
    final linkElements = feed.findElements('link');
    xml.XmlElement? linkElement;
    for (final el in linkElements) {
      if (el.getAttribute('rel') == 'alternate') {
        linkElement = el;
        break;
      }
    }
    linkElement ??= linkElements.firstOrNull;
    final link = linkElement?.getAttribute('href') ?? '';

    final items = <ParsedItem>[];
    for (final entry in feed.findElements('entry')) {
      final entryLinks = entry.findElements('link');
      xml.XmlElement? entryLink;
      for (final el in entryLinks) {
        if (el.getAttribute('rel') == 'alternate') {
          entryLink = el;
          break;
        }
      }
      entryLink ??= entryLinks.firstOrNull;

      final authorElement = entry
          .findAllElements('author')
          .expand((a) => a.findAllElements('name'))
          .firstOrNull;

      final parsedItem = ParsedItem(
        title: _getText(entry.findElements('title').firstOrNull),
        link: entryLink?.getAttribute('href') ?? '',
        description: _getText(entry.findElements('summary').firstOrNull),
        content: _getText(entry.findElements('content').firstOrNull),
        pubDate:
            _getText(_firstOf(entry, ['published', 'updated'])),
        author: _getText(authorElement),
        guid: _getText(entry.findElements('id').firstOrNull) ??
            (entryLink?.getAttribute('href') ?? ''),
        imageUrl: _extractImageUrl(entry),
      );
      items.add(parsedItem);
    }

    return ParsedFeed(
      title: title,
      description: subtitle,
      link: link,
      items: items,
    );
  }

  xml.XmlElement? _firstOf(xml.XmlElement parent, List<String> names) {
    for (final name in names) {
      final parts = name.split(':');
      final localName = parts.last;
      final elements = parent.findAllElements(localName);
      if (elements.isNotEmpty) {
        return elements.first;
      }
    }
    return null;
  }

  String _getText(xml.XmlElement? element) {
    if (element == null) return '';
    return element.text.trim();
  }

  xml.XmlElement? _findElementWithLocalName(
    xml.XmlElement parent,
    String name,
  ) {
    for (final element in parent.descendants.whereType<xml.XmlElement>()) {
      if (element.name.local == name) {
        return element;
      }
    }
    return null;
  }

  String? _extractImageUrl(xml.XmlElement item) {
    final mediaContent = _findElementWithLocalName(item, 'content');
    if (mediaContent != null) {
      final url = mediaContent.getAttribute('url');
      if (url != null && url.isNotEmpty) return url;
    }

    final mediaThumbnail = _findElementWithLocalName(item, 'thumbnail');
    if (mediaThumbnail != null) {
      final url = mediaThumbnail.getAttribute('url');
      if (url != null && url.isNotEmpty) return url;
    }

    final enclosure = item.findElements('enclosure').firstOrNull;
    if (enclosure != null) {
      final type = enclosure.getAttribute('type') ?? '';
      if (type.startsWith('image/')) {
        final url = enclosure.getAttribute('url');
        if (url != null && url.isNotEmpty) return url;
      }
    }

    final descriptionEl = item.findElements('description').firstOrNull;
    final contentEl =
        _firstOf(item, ['content:encoded', 'content']) ?? descriptionEl;
    final htmlContent = contentEl != null ? contentEl.text : '';
    if (htmlContent.isNotEmpty) {
      final match = RegExp(
        r'<img[^>]+src=["\']([^"\']+)["\']',
        caseSensitive: false,
      ).firstMatch(htmlContent);
      if (match != null && match.groupCount >= 1) {
        return match.group(1);
      }
    }

    return null;
  }

  DateTime? _parseDate(String? value) {
    if (value == null || value.isEmpty) return null;
    try {
      return DateTime.parse(value);
    } catch (_) {
      return null;
    }
  }

  Future<int> refreshFeed(RssFeed feed) async {
    final parsedFeed = await fetchFeed(feed.url);

    final updatedFeed = feed.copyWith(
      title: parsedFeed.title.isNotEmpty ? parsedFeed.title : feed.title,
      description: parsedFeed.description ?? feed.description,
      link: parsedFeed.link ?? feed.link,
      lastFetched: DateTime.now(),
    );
    await dbService.updateFeed(updatedFeed);

    final existingArticles = await dbService.getArticlesByFeed(feed.id);
    final existingGuids = existingArticles.map((a) => a.guid).toSet();

    final newArticles = <RssArticle>[];
    for (final item in parsedFeed.items) {
      final key = item.guid ?? item.link;
      if (key.isEmpty || existingGuids.contains(key)) {
        continue;
      }
      final article = RssArticle.create(
        feedId: feed.id,
        title: item.title,
        link: item.link,
        description: item.description,
        content: item.content,
        pubDate: _parseDate(item.pubDate),
        author: item.author,
        guid: key,
        imageUrl: item.imageUrl,
      );
      newArticles.add(article);
    }

    if (newArticles.isNotEmpty) {
      await dbService.addArticles(newArticles);
    }

    return newArticles.length;
  }

  Future<RefreshAllResult> refreshAllFeeds(List<RssFeed> feeds) async {
    var total = 0;
    final errors = <String>[];

    for (final feed in feeds) {
      try {
        final count = await refreshFeed(feed);
        total += count;
      } catch (e) {
        errors.add('${feed.title}: $e');
      }
    }

    return RefreshAllResult(total: total, errors: errors);
  }
}

extension _FirstOrNullExtension<E> on Iterable<E> {
  E? get firstOrNull {
    final iterator = this.iterator;
    if (!iterator.moveNext()) return null;
    return iterator.current;
  }
}
